Further SpringBoot improvements

zambrovski opened the pull request:
Hi @zdenek-jonas 

I included several improvements in this PR, including your work on #129 

- fix #130 
- fix #133 
- fix #129 

## SpringBoot Inaccuracies (#129)

I rewrote the configuration of the SpringBoot, by applying the following ideas. The imperative code for creation of the `EmbeddedStoreFoundation` and `EmbeddedStoreManager` are moved to special factory classes. The old `StorageManagerProvider` is now deprecated (and delegates to the factories), since it is not a provider but rather a factory (it creates a new instance) and I performed the separation, because of the violation of the SRP.

Having said that, I created bean factories for all basic facilities in the main `EclipseStoreSpringBoot` (auto-config). These are for now:
- `EclipseStoreConfigConverter`
- `ClassLoaderProvider`
- `EmbeddedStorageFoundationFactory`
- `EmbeddedStorageManagerFactory`

In addition, I created the `DefaultEclipseStoreConfiguration` which will instantiate and create an `EmbeddedStorageManager` based on properties if there is no other defined. This auto-configuration can be easily excluded, or one might provide a configuration creating own `EmbeddedStorageManager` just before it, preventing the default to be created.

By doing so, we avoid `@Lazy`, `@Primary` and other Spring Black Magic making developer business hard.

I slightly changed the SpringTest setup introducing a real SpringBoot "application", this makes configs a lot easier.

## SpringBoot Web MVC REST Service (#130)

I created a new Maven module `storage/rest/service-springboot` providing a Spring Web MVC implementation of the REST controller. This modules relies on standard Web MVC stuff like Jackson and adopts the `StorageRestAdapter` output in a Spring way. Currently, I bind this controller once. In the future, I want to detect multiple configured `EmbeddedStorageManager` instances and bind it several times to them. I assumed that the issue #135 resolves towards `application/json` as content type.

I introduced configuration for the SpringBoot REST Service. It includes two values:
- `enabled` (defaults to `false`) activates the creation of the REST controller. By doing so, you can't occasionally put the JAR on the class path and get an unsecured endpoint deployed to production, just by magic Spring Auto configuration. Now you have to explicitly activate the REST service in the properties.
- `base-url` (defaults to `/store-data`, following the same `base-url` as the Spark Java instance name) is the base request-mapping of the REST service controller. It can be changed via property and then put under some other path - for example because of some security considerations.
- If you are using the default configuration (for one storage manager) it will expose its data under `/store-data/default`. If you bind multiple storage managers, it will expose their data under the paths matching the storage manager bean names (if you provide `first` and `second` you'll get `/store-data/first` and `/store-data/second`.


## Separation of Client App in Implementaton and Assembly (#133)

I created a new module `storage/rest/client-app-standalone-assembly` and moved all the SpringBoot application stuff into it, leaving a simple Configuration class in the `client-app` module. By doing so, it is now possible to use the `client-app` module as implementation of the console and embed it to any SpringBoot application.

## Small path fix (no issue)
 
I realized that in `client-app` the `UIUtils` class had a method creating a path to the static resources. The method was missing a leading "/". This doesn't matter if the app is bound to "/" but breaks if the Vaadin app is bound to any other path, since the static resources are fixed in the JAR.

## Created the console

A new additional module `integrations/spring-boot3-console` is created. This is a small module for unifying the Spring REST Service and Client-App and provides some minimal configuration so far. The idea is that if you want to have the ability to use the Vaadin Console in an existing SpringBoot ap, you would include the `spring-boot3-console` and use it.

## Created `integrations/itest` 

This module is needed in order to cope with Maven dependency scopes. If we build libraries which use `provided` dependencies, there is no way to run Spring Boot ITests in which those dependencies are there. If your tests only rely on spring-context, it is possible, for web, data and other aspects an external module (with dependencies) is required. So this module should include Spring Boot Tests with more complex setup. I put the MockMVC tests of the REST service inside for the moment.

## Minor changes to Spring Boot Example

- activated the SpringBoot Console in SpringBoot Example.
- added swagger-ui to provide an easy REST client for the users.


## Documentation

I added/adjusted documentation for:

- SpringBoot Integration
- REST Service Spring Boot
- REST Client App (Standalone mode)
- REST Client App as SpringBoot Console as part of SpringBoot Integration
- Example README for SpringBoot

zambrovski commented:
@zdenek-jonas my idea is to start slowly, understand how do you work, what do you like and what not and then provide more interesting (but rather complex) implementations like dynamic registration of multiple store managers in the statistics...

zdenek-jonas commented:
@zambrovski 
I will just push small fixes, the rest will be done with comments.  Good job!

zdenek-jonas commented:
TODO: Before Merge

- [ ] Docu. There is a chapter about the standalone app, this was changed, fix docu. `docs/modules/storage/pages/rest-interface/client-gui.adoc`

zambrovski commented:
I could supply some docs indeed. And I would rebase, as soon as you merge in the fixes of #129

zdenek-jonas commented:
> I could supply some docs indeed. And I would rebase, as soon as you merge in the fixes of #129

Could you help me with this issue? 
Especially point 4 and 5?

zdenek-jonas commented:
@zambrovski 
From a security standpoint, an idea struck me. By simply adding the dependency for spring-boot-console, the application will expose a Rest API, allowing data retrieval. While I can configure to disable the console and UI, the Rest API remains active. I can envision a scenario where this could inadvertently make its way into production. Perhaps it would be safer to introduce configuration for the Rest API, defaulting to false. This would require explicit configuration in the file to enable it. Originally designed for developers to inspect data, this Rest API lacks any implemented security measures, etc.

zdenek-jonas commented:
We have to internally discuss, if the enable this view as default is a good idea from security view. Maybe default disable and just enable it in the configuration. When somebody add this dependency into pom.xml, it will enable by default.

zambrovski commented:
Full agree, I'll introduce the configuration disabled by default.

zambrovski commented:
I introduced configuration for the SpringBoot REST Service.

It includes two values:
- `enabled` (defaults to `false`) activates the creation of the REST controller. By doing so, you can't occasionally put the JAR on the class path and get an unsecured endpoint deployed to production, just by magic Spring autoconfiguration. Now you have to explicitly activate the REST service in the properties.
- `base-url` (defaults to `store-data`, following the same base-url as the Spark Java instance name) is the base request-mapping of the REST service controller. It can be changed via property and then put under some other path - for example because of some security considerations.

I adjusted the docs (took me a moment to get Antora running in Docker) and added the section about SpringBoot REST service and the changes of the standalone app. Currently, the Console in SpringBoot mode is not included, probably it is a good idea to include it into the docs about the SpringBoot integration. 

I will merge the changes you made regarding #129 into this branch and create one PR on all this.

zambrovski commented:
I updated the PR description above and included the items I solved from #129 
The only open one is Issue 5 of it. Let us split it up in a separate issue and discuss there.

zambrovski commented:
@zdenek-jonas I'm ready from my side. Feel free to review please...

zdenek-jonas commented:
I am currently reviewing this Pull Request and, in principle, I haven't found any major issues. There are only minor points that I will note down as I come across them, rather than waiting until the end.  

Regarding the **_client-app-standalone-assembly_**:  
We have temporarily removed the module-info.java file due to some build issues that we were unable to resolve at this time. However, we plan to reintroduce it later. I would like to know if it's possible to modify the package of this module to make it unique. I believe it is, but I want to confirm. This is the reason behind our package structure in the project.

zambrovski commented:
I'll have a look on it. 
I think the docs is then not precise enough. What it should state is:

- If you have a single storage configured from the application properties, this works out of the box.
- If you need to do some manual modification on the foundation or on the storage manager, you can do this, redefining the storage manager bean factory (this is what you do in your component, but the correct way would be to create an own class annotated with `@Configuration` and do it there)
- If you have multiple storages - then the same as in 2 but using qualified properties and qualified storage manager. 

I also envision an easy possibility to bind the REST storage to multiple managers. I'll create this small addition on monday.

zambrovski commented:
I'll create a second (advanced) example demonstrating the usage of two storage managers. I studied your example and will base on that for it...

zambrovski commented:
@zdenek-jonas I pushed latest changes with a second example and advanced REST service integration capable to handle multiple storage managers. Please review.

zdenek-jonas commented:
We have already added the module-info in the main module. Next, I will include the module info in the new modules in order to merge them and ensure the build checks are functioning properly.

zdenek-jonas commented:
@zambrovski Could you rebase this project from main. The version 1.2.0-Snapshot has been released, actual is 1.3.0-Snapshot. Or I can do it. But this is your repository, I want to just clear, before I make such changes in your repository.

zambrovski commented:
Feel free to rebase…
I‘ll be happy if you take this over.



Simon Zambrovski
Senior Consultant
Holisticon AG

Jürgen-Töpfer-Str. 44
22763 Hamburg
Mobile: +49 176 1616 9010
Office: +49 40 6094 430-0
Fax: +49 40 6094 430-30
E-Mail: ***@***.***
Twitter: @zambrovski
www.holisticon.de

Amtsgericht Hamburg, HRB 107396
Vorstand: Oliver Ihns, Dierk Harbeck
Aufsichtsrat: Sune Nilsson (Vorsitz)


Am 21.02.2024 um 21:55 schrieb Zdenek Jonas ***@***.***>:

﻿

@zambrovski<https://github.com/zambrovski> Could you rebase this project from main. The version 1.2.0-Snapshot has been released, actual is 1.3.0-Snapshot. Or I can do it. But this is your repository, I want to just clear, before I make such changes in your repository.

—
Reply to this email directly, view it on GitHub<https://github.com/eclipse-store/store/pull/136#issuecomment-1957898940>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAFEK2HD2ORU3EGJPAANDBTYUZNKHAVCNFSM6AAAAABC4QFZTGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNJXHA4TQOJUGA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

zdenek-jonas commented on docs/modules/storage/pages/rest-interface/client-gui.adoc:
```diff
@@ -6,29 +6,29 @@ It is a runnable jar which starts a simple web server which then can be accessed
 [source, shell, title="CLI"]
 ----
-mvn dependency:get -Dartifact=org.eclipse.store:storage-restclient-app:{maven-version}:jar -Dtransitive=false -Ddest=storage.restclient.app-{maven-version}.jar
+mvn dependency:get -Dartifact=org.eclipse.store:storage-restclient-app-stanalone:{maven-version}:jar -Dtransitive=false -Ddest=storage.restclient.app-{maven-version}.jar
```
fix

fh-ms commented on storage/embedded/pom.xml:
```diff
@@ -29,6 +29,10 @@
 			<groupId>org.slf4j</groupId>
 			<artifactId>slf4j-api</artifactId>
 		</dependency>
+		<dependency>
```
Double dependency
